### PR TITLE
Add solution for assignment3

### PR DESCRIPTION
**Pull request by nfadeeva**

### CODE REVIEW

File: nfadeeva/assignment3/dictionary.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+class Node:
+    def __init__(self):
+        self.children = collections.defaultdict(Node)
+        self.is_word = False
+
+
+class Dictionary:
+    """
+    Dictionary class that looks like a trie
```
Comment (ariava) [inline nfadeeva/assignment3/dictionary.py:13]: Super cool!

File: nfadeeva/assignment3/tests.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import unittest
+
+
+class Test(unittest.TestCase):
+
+    def setUp(self):
+        self.d = Dictionary(['CAR', 'CAT', 'CARD', 'CART'])
+
+    # tests for the Dictionary class
+    def test_is_word_true(self):
```
Comment (ariava) [inline nfadeeva/assignment3/tests.py]: You can have more descriptive names, e.g. "test_word_is_in_dictionary" or "test_prefix_is_in_dictionary"

File: nfadeeva/assignment3/search_words.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+from typing import List
+
+
+class Grid:
+    def __init__(self, cells: List[List[str]]):
+        self.cells = cells
+        self.r = len(cells)
```
Comment (ariava) [inline nfadeeva/assignment3/search_words.py]: Please use clearer names, e.g. "rows" and "columns"

File: nfadeeva/assignment3/search_words.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    def neighbours(self, i, j):
+        """
+        Get all valid neighbours for the given cell
+        """
+
+        res = []
+        for x in range(i - 1, i + 2):
+            for y in range(j - 1, j + 2):
+                if x == i and y == j:
```
Comment (ariava) [inline nfadeeva/assignment3/search_words.py]: Can you merge the two ifs?

File: nfadeeva/assignment3/search_words.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    if used is None:
+        used = set()
+    used.add((i, j))
+    prefix += cells[i][j]
+    if dictionary.is_word(prefix):
+        words.add(prefix)
+
+    # go through all valid neighbours of the given cell to collect more words
+    for x, y in grid.neighbours(i, j):
+        if (x, y) not in used and dictionary.is_prefix(prefix + cells[x][y]):
```
Comment (ariava) [inline nfadeeva/assignment3/search_words.py]: Optimization: can you skip iterating on the neighbors if not dictionary.is_prefix(prefix)?

File: nfadeeva/assignment3/dictionary.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    """
+
+    def __init__(self, data: List[str]):
```
Comment (ariava) [inline nfadeeva/assignment3/dictionary.py:16]: Good job using annotations to clarify what you expect as an input!

File: nfadeeva/assignment3/dictionary.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        """
+        :param data: list of words
+        :type data: List[str]
+        """
+
+        self.root = Node()
+        for word in data:
+            self.put(word)
+
+    def put(self, word):
```
Comment (ariava) [inline nfadeeva/assignment3/dictionary.py]: nit: use a different name. Conventionally, a method called "put" usually decreases a reference count on an object, it means that some user is no longer using that object. Here you are doing exactly the opposit.
You could call the method e.g. "add".

**Review by ariava — CHANGES_REQUESTED** [review]
Great job! Please have a look at my comments.

File: nfadeeva/assignment3/search_words.py (hunk 3 above)
Comment (nfadeeva) [reply nfadeeva/assignment3/search_words.py]: @ariava  I've already fixed my code according to your comments, you can check it:)

**Review by ariava — APPROVED** [review]
Good job!

**Pull request merged**